cmd: document updated `exp run --reset` behavior (#2286)
iterative/dvc.org: content/docs/command-reference/exp/run.md

> To do so, first mark stage `outs` with `checkpoint: true` in `dvc.yaml`. Having
> at least a checkpoint <abbr>output</abbr> is needed so that the experiment can

Reviewer: ```suggestion
To do so, first mark stage `outs` with `checkpoint: true` in `dvc.yaml`.
At least one checkpoint <abbr>output</abbr> is needed so that the experiment can
```

Reviewer: Minor style suggestion

Reviewer 2: Applied!

later continue from that output's last <abbr>cached</abbr> state.

> ⚠️ Using the `checkpoint` field in `dvc.yaml` is only compatibly with

Author: ```suggestion
⚠️ Using the `checkpoint` field in `dvc.yaml` is only compatible with
```

typo here

`dvc exp run`, `dvc repro` will abort if any stage contains it.

Then, in your code either call the `dvc.api.make_checkpoint()` function
(Python), or write a signal file (any programming language) following the same
steps as `make_checkpoint()` — please refer to its reference for details.

You can now use `dvc exp run` to begin the experiment. This removes any
`checkpoint` outputs before running the experiment (regardless of whether they
have cached versions). When the process finishes or gets interrupted (e.g. with
Ctrl + `C`), DVC will [apply](/doc/command-reference/exp/apply) the last
checkpoint to the <abbr>workspace</abbr> (overwriting any further changes done

> by the stage).

Author: I think wording it this way might end up confusing users?

>  This removes any `checkpoint` outputs before running the experiment (regardless of whether they have cached versions).

`checkpoint` outs are only removed before running the experiment if `--reset` is specified. In a normal `dvc exp run` with no other arguments, we continue the experiment using the most recent version of that output.

If there is no previous run at all and the output has never existed before, then yes `dvc exp run` will do the same thing as `dvc exp run --reset`, but only in that specific case.

Reviewer: Can this sentence get deleted? There's nothing special about what happens when you begin a checkpoint experiment. It's only different after at least one checkpoint has been generated.

Reviewer 2: > There's nothing special about what happens when you begin a checkpoint experiment

What about what Peter just mentioned @dberenbaum? "If there is no previous run at all and the output has never existed before, then yes dvc exp run will do the same thing as dvc exp run --reset" 🙂

Reviewer 2: > wording it this way might end up confusing users?

@pmrowla good point but maybe the problem is that the behavior can be a bit confusing per-se. I'll see if there's a more explicit way to express this but it already states "use `dvc exp run` **to begin** the experiment. This removes any `checkpoint` outputs" (it specifies "to begin").

Author: That's not really a special case though, and even in that case nothing is being removed (since there is no initial data to even remove in the first place)

Reviewer 2: Nvmd I think my latest update should make it very clear. It's getting a little long/convoluted but at least it's complete and correct (I think). See 9e1a7e94.

Reviewer 2: > even in that case nothing is being removed (since there is no initial data to even remove in the first place

What if you just changed regular outputs into checkpoint outputs in the last commit? (to migrate from regular pipeline to a checkpoint experiment)

Reviewer: So the updated sentence is:

```
On this very first use,
any `checkpoint` outputs are deleted before running the experiment (regardless
of whether they have <abbr>cached</abbr> versions). 
```

> It's getting a little long/convoluted but at least it's complete and correct (I think).

Agreed on all counts 😄 . I think I get your point now that it's unclear whether an initial checkpoint run treats existing outs as checkpoints or not. Ideally, this is called out as an edge case since it might add confusion for the typical case where no outs exist yet, but 🤷 .

Reviewer 2: > the typical case where no outs exist yet

OK if its an edge case we can leave it undocumented. Always good to simplify docs. Removed that sentence for now... Not sure I fully understand all the possible scenarios but will try to QA separately.

> checkpoint to the <abbr>workspace</abbr> (overwriting any further changes done
> by the stage).

Reviewer: ```suggestion
checkpoint to the <abbr>workspace</abbr> (overwriting any further changes done
by the stage if the process was interrupted).
```
Alternatively:
```
When the process finishes, DVC will [apply](/doc/command-reference/exp/apply) the last
checkpoint to the <abbr>workspace</abbr>. If the process gets interrupted  (e.g. with
Ctrl + `C`), DVC will overwrite any changes done by the stage after the last checkpoint.
```
I find "overwriting any further changes" confusing unless it's specifically tied to the process being interrupted. If the process isn't interrupted, I think a final checkpoint is automatically generated to save the final outcome, so it's not overwriting further changes (@pmrowla can you confirm if that's correct?).

Author: Yeah, there's an additional commit generated at the end of the stage (assuming anything changed since the last time the user called `make_checkpoint`), and that is what gets applied. Even if the process is interrupted, we don't overwrite anything in the user's workspace.

Reviewer: So should `(overwriting any further changes done by the stage)` get dropped?

Reviewer 2: > there's an additional commit generated at the end of the stage

Ah this I didn't realize. OK I rewrote this based on that and Dave's suggestions (thanks) for now.

> should (overwriting any further changes done by the stage) get dropped?

Woah wo why did I think that? A bit confused on this, so a final checkpoint that reflects the resulting workspace "is applied" (what is there to apply?) but the last checkpoint for interrupted runs is NOT applied?

Reviewer: Here's the latest language for reference:

```
If the process gets
interrupted (e.g. with Ctrl + `C`), DVC will
[apply](/doc/command-reference/exp/apply) the last checkpoint to the
<abbr>workspace</abbr> (overwriting any further changes).
```

Pending confirmation from @pmrowla, maybe we should remove any reference to `apply` or the workspace if nothing gets overwritten? Maybe this sentence isn't needed at all? Do we feel it necessary to call out process interruption to make explicit that this is an expected workflow? If so, maybe something like:

```
If the process gets
interrupted (e.g. with Ctrl + `C`), DVC will
keep all checkpoints recorded before interruption.
```

Reviewer 2: OK. Removed `apply` and workspace references for now per Peter's comment above but it would def. be great to confirm!

`dvc exp run`, `dvc repro` will abort if any stage contains it.

Then, in your code either call the `dvc.api.make_checkpoint()` function
(Python), or write a signal file (any programming language) following the same
steps as `make_checkpoint()` — please refer to its reference for details.

You can now use `dvc exp run` to begin the experiment. On this very first use,
any `checkpoint` outputs are deleted before running the experiment (regardless
of whether they have cached versions). When the process finishes or gets
interrupted (e.g. with Ctrl + `C`), DVC will
[apply](/doc/command-reference/exp/apply) the last checkpoint to the
<abbr>workspace</abbr> (overwriting any further changes done by the stage).

Following uses of `dvc exp run` will continue from this point (using the latest
cached versions of all outputs). You can add a `--rev` to continue from a
previous checkpoint instead (list them with `dvc exp show`). Or use `--reset` to
start over (discards previous checkpoints and deletes `checkpoint` outputs, like
the first `dvc exp run`) — useful for re-training ML models, for example.

> Note that after resetting checkpoint experiments, you may want to use `dvc gc`

> > to clean up unnecessary data in the cache.

Author: I don't think this note is really necessary, but I suppose it doesn't hurt anything either.

Reviewer 2: Can't resetting checkpoints several times fill up the cache pretty quickly?

Author: Sure, but that just comes with running a lot of experiments in general, it's not really specific to checkpoints.

Reviewer 2: 🤺 you're right. Removed from here but made a slightly more prominent note about this in the general Description.

<abbr>workspace</abbr> (overwriting any further changes done by the stage).

`dvc exp run` again will continue from this point (useful for interrupted runs).
Use `--reset` to reset (remove) any existing `checkpoint` outputs in your

> workspace before running the experiment. Note that `--reset` overrides any

Reviewer 2: Sounds like "use reset to reset". We could just keep "to remove" I think. But...

> (remove) any existing `checkpoint` outputs

A little low level. What does `--reset` do from the user perspective. Does it not roll back to `HEAD` anymore?

Reviewer 2: To be clearer, here are the steps implied in this section (checkpoints workflow with reset):

0. enable checkpoints in the stage ✅
1. update code to write signal files ✅
2. `exo run` ✅ (and possibly interrupt)
3. \> `exp run --reset` < - We are here

Why would we want to reset? Again, to restart the experiment, right? (Go back to step 2.)

Author: For `checkpoint` outs, `reset` removes the specified outs entirely (so that the specified intermediate model can be retrained). It used to roll back to `HEAD`, but this was changed (as of the core PR for these docs).

see https://github.com/iterative/dvc/issues/5553 for the product-side discussion on why this change was made

Reviewer 2: I see, thanks for the context.

> For checkpoint outs

Is there another use case for `--reset`?

Reviewer 2: How about

```md
Use `--reset` to remove `checkpoint` outputs completely to restart the
experiment from scratch regardless of any cached contents.
```

But should we move `exp run --reset` up in the steps or say it's an alternative to plain `exp run` from the get-go?

0, 1, (same as above)
2a. `exp run` - runs experiment from `HEAD` state
2b. `exp run --reset` - runs experiment from scratch

Author: If it's a new stage with new `checkpoint` outputs, both `2a` and `2b` are the same thing here (since there would not be any HEAD `dvc.lock` entries for the checkpoint outs).

Reviewer 2: OK so the first `exp run` of a checkpoint pipeline is the same as `exp run --reset`, got it. So yeah this whole section needs a bit of work to include all these key points. Please address my other comments/suggestions when you can and lmk to take this over a some final reorg of ideas (I'll try it in the meantime to confirm I got it 100% now).

Reviewer 2: One more Q @pmrowla. `exp run --reset` still deletes the checkpoints also, right? Meaning the Git commits (I'm guessing the cache is left alone).

Author: The experiment ref for the previously existing checkpoint run is what gets "deleted". In practice, as long as the user's stage is deterministic the old existing is just moved to point to the new run.

Git commits are left alone, git will garbage collect them on its own once they are no longer pointed to by any references.

Reviewer 2: Noted.

> existing `dvc.lock` entries for `checkpoint` outputs. Alternatively, you can use

Reviewer 2: > `--reset` overrides any existing `dvc.lock` entries for `checkpoint` outputs

I'm not getting this note, what are we trying to warn about?

Author: If your `dvc.lock` file is git-committed and the HEAD version of `dvc.lock` has a hash for the specified `checkpoint` out, that hash will be ignored on `--reset`, and the out will be completely removed before reproducing the pipeline (rather than being checked out to HEAD's hash).

Reviewer: To put this in user terms, if you save a model at each checkpoint so that its weights get loaded as the starting point for the next training epoch, `--reset` will drop the model so that training will start without any existing model weights.

> Use `--reset` to reset (remove) any existing `checkpoint` outputs in your
> workspace before running the experiment. Note that `--reset` overrides any
> existing `dvc.lock` entries for `checkpoint` outputs. Alternatively, you can use
> `--rev` to continue from a specific (previous) checkpoint.

Reviewer 2: Consolidated suggestion for this part:
```suggestion
Use `--reset` to remove `checkpoint` outputs completely in order to restart an
experiment from scratch, regardless of any cached contents (`dvc.lock` is
disregarded) — useful for re-training ML models, for example. Alternatively, you
can use `--rev` to continue from a specific (previous) checkpoint.
```

Reviewer 2: But also the earlier sentence (same section) "Having at least a checkpoint output is needed so that the experiment can later restart based on that output's last cached state." should probably be rephrased.

is actually run. Instead, the experiment is put in a wait-list for later
execution. `dvc exp show` will mark queued experiments with an asterisk `*`.

> Note that for experiments which use checkpoints, queuing an experiment implies
> `--reset` unless an existing checkpoint to resume is explicitly provided via
> `--rev`.
> 
> Use `dvc exp run --run-all` to process this queue.
> 
> Adding `-j` (`--jobs`), experiment queues can be run in parallel for better
> performance. This creates a temporary workspace copy for each subprocess (in

Reviewer 2: ```suggestion
> Note that queuing an experiment that uses checkpoints implies `--reset`
> unless `--rev` is used.

Use `dvc exp run --run-all` to process the queue. Adding `-j` (`--jobs`),
experiment queues can be run in parallel for better performance. This creates a
temporary workspace copy for each subprocess (in
```

The 2nd p needs wrapping with the remaining line but I couldn't include it in the suggestion.

- `--queue` - place this experiment at the end of a line for future execution,
  but do not actually run it yet. Use `dvc exp run --run-all` to process the

>   queue. For checkpoint experiments, this implies `--reset` unless `--rev` is
>   provided.

Reviewer 2: Quick copy edit (will commit):
```suggestion
  queue. For checkpoint experiments, this implies `--reset` unless a `--rev` is
  provided.
```

  checkpoint name or hash (`commit`). This is needed for example to resume
  experiments from `--queue` or `--temp` runs.

- `--reset` - remove `checkpoint` outputs before running this experiment
  (regardless of `dvc.lock`). Implies `--force`, so that cached checkpoint

>   results are regenerated. Useful for ML model re-training

Reviewer: ```suggestion
  results are regenerated. Useful for ML model re-training.
```

  checkpoint name or hash (`commit`). This is needed for example to resume
  experiments from `--queue` or `--temp` runs.

> - `--reset` - reset any checkpoint outputs before running this experiment.
>   Implies `--force`, so that cached checkpoint results are regenerated.

Reviewer 2: ```suggestion
- `--reset` - remove `checkpoint` outputs before running this experiment
  (regardless of `dvc.lock` or their cached contents). Implies `--force`, so
  that cached checkpoint results are regenerated. Useful for ML model
  re-training.
```
